### PR TITLE
Enable PowerSGD on XLM-R cross-lingual MLM task

### PR DESCRIPTION
**Pull request by SciPioneer**
No need to review this PR. No plan to land the change.

This PR registers a PowerSGD communication hook when a DDP model is created.

Please try it on FP32 input data.

### CODE REVIEW

**Comment by zhangguanheng66** [comment]
Run the case and got the follow error message
```
(base) zhangguanheng@devfair0197:BERT$ srun --label --ntasks-per-node=1 --time=4000 --mem-per-cpu=5120 --gres=gpu:8 --cpus-per-task 80 --nodes=1 --pty python cross_lingual_mlm_task.py --num_lines 10000 --world_size 8 --pipeline_mode DDP --epochs 12 --lr 0.01 --bptt 96 --log-interval 50
srun: job 37541787 queued and waiting for resources
srun: job 37541787 has been allocated resources
0: Traceback (most recent call last):
0:   File "cross_lingual_mlm_task.py", line 380, in <module>
0:     run_demo(run_ddp, run_main, args)
0:   File "/private/home/zhangguanheng/tmp/PyTorch/text/examples/BERT/utils.py", line 26, in run_demo
0:     join=True)
0:   File "/private/home/zhangguanheng/tmp/PyTorch/pytorch/torch/multiprocessing/spawn.py", line 230, in spawn
0:     return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')
0:   File "/private/home/zhangguanheng/tmp/PyTorch/pytorch/torch/multiprocessing/spawn.py", line 188, in start_processes
0:     while not context.join():
0:   File "/private/home/zhangguanheng/tmp/PyTorch/pytorch/torch/multiprocessing/spawn.py", line 150, in join
0:     raise ProcessRaisedException(msg, error_index, failed_process.pid)
0: torch.multiprocessing.spawn.ProcessRaisedException:
0:
0: -- Process 1 terminated with the following error:
0: Traceback (most recent call last):
0:   File "/private/home/zhangguanheng/tmp/PyTorch/pytorch/torch/multiprocessing/spawn.py", line 59, in _wrap
0:     fn(i, *args)
0:   File "/private/home/zhangguanheng/tmp/PyTorch/text/examples/BERT/utils.py", line 31, in run_ddp
0:     main_fn(args, rank)
0:   File "/private/home/zhangguanheng/tmp/PyTorch/text/examples/BERT/cross_lingual_mlm_task.py", line 155, in run_main
0:     self.register_comm_hook(state, powerSGD.powerSGD_hook)
0: NameError: name 'self' is not defined
0:
srun: error: learnfair1730: task 0: Exited with exit code 1
srun: launch/slurm: _step_signal: Terminating StepId=37541787.0
```

**Comment by SciPioneer** [comment]
> Run the case and got the follow error message
> 
> ```
> (base) zhangguanheng@devfair0197:BERT$ srun --label --ntasks-per-node=1 --time=4000 --mem-per-cpu=5120 --gres=gpu:8 --cpus-per-task 80 --nodes=1 --pty python cross_lingual_mlm_task.py --num_lines 10000 --world_size 8 --pipeline_mode DDP --epochs 12 --lr 0.01 --bptt 96 --log-interval 50
> srun: job 37541787 queued and waiting for resources
> srun: job 37541787 has been allocated resources
> 0: Traceback (most recent call last):
> 0:   File "cross_lingual_mlm_task.py", line 380, in <module>
> 0:     run_demo(run_ddp, run_main, args)
> 0:   File "/private/home/zhangguanheng/tmp/PyTorch/text/examples/BERT/utils.py", line 26, in run_demo
> 0:     join=True)
> 0:   File "/private/home/zhangguanheng/tmp/PyTorch/pytorch/torch/multiprocessing/spawn.py", line 230, in spawn
> 0:     return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')
> 0:   File "/private/home/zhangguanheng/tmp/PyTorch/pytorch/torch/multiprocessing/spawn.py", line 188, in start_processes
> 0:     while not context.join():
> 0:   File "/private/home/zhangguanheng/tmp/PyTorch/pytorch/torch/multiprocessing/spawn.py", line 150, in join
> 0:     raise ProcessRaisedException(msg, error_index, failed_process.pid)
> 0: torch.multiprocessing.spawn.ProcessRaisedException:
> 0:
> 0: -- Process 1 terminated with the following error:
> 0: Traceback (most recent call last):
> 0:   File "/private/home/zhangguanheng/tmp/PyTorch/pytorch/torch/multiprocessing/spawn.py", line 59, in _wrap
> 0:     fn(i, *args)
> 0:   File "/private/home/zhangguanheng/tmp/PyTorch/text/examples/BERT/utils.py", line 31, in run_ddp
> 0:     main_fn(args, rank)
> 0:   File "/private/home/zhangguanheng/tmp/PyTorch/text/examples/BERT/cross_lingual_mlm_task.py", line 155, in run_main
> 0:     self.register_comm_hook(state, powerSGD.powerSGD_hook)
> 0: NameError: name 'self' is not defined
> 0:
> srun: error: learnfair1730: task 0: Exited with exit code 1
> srun: launch/slurm: _step_signal: Terminating StepId=37541787.0
> ```

Thanks! Fixed it. Please make sure you are using NCCL backend.

**Comment by zhangguanheng66** [comment]
> > Run the case and got the follow error message
> > ```
> > (base) zhangguanheng@devfair0197:BERT$ srun --label --ntasks-per-node=1 --time=4000 --mem-per-cpu=5120 --gres=gpu:8 --cpus-per-task 80 --nodes=1 --pty python cross_lingual_mlm_task.py --num_lines 10000 --world_size 8 --pipeline_mode DDP --epochs 12 --lr 0.01 --bptt 96 --log-interval 50
> > srun: job 37541787 queued and waiting for resources
> > srun: job 37541787 has been allocated resources
> > 0: Traceback (most recent call last):
> > 0:   File "cross_lingual_mlm_task.py", line 380, in <module>
> > 0:     run_demo(run_ddp, run_main, args)
> > 0:   File "/private/home/zhangguanheng/tmp/PyTorch/text/examples/BERT/utils.py", line 26, in run_demo
> > 0:     join=True)
> > 0:   File "/private/home/zhangguanheng/tmp/PyTorch/pytorch/torch/multiprocessing/spawn.py", line 230, in spawn
> > 0:     return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')
> > 0:   File "/private/home/zhangguanheng/tmp/PyTorch/pytorch/torch/multiprocessing/spawn.py", line 188, in start_processes
> > 0:     while not context.join():
> > 0:   File "/private/home/zhangguanheng/tmp/PyTorch/pytorch/torch/multiprocessing/spawn.py", line 150, in join
> > 0:     raise ProcessRaisedException(msg, error_index, failed_process.pid)
> > 0: torch.multiprocessing.spawn.ProcessRaisedException:
> > 0:
> > 0: -- Process 1 terminated with the following error:
> > 0: Traceback (most recent call last):
> > 0:   File "/private/home/zhangguanheng/tmp/PyTorch/pytorch/torch/multiprocessing/spawn.py", line 59, in _wrap
> > 0:     fn(i, *args)
> > 0:   File "/private/home/zhangguanheng/tmp/PyTorch/text/examples/BERT/utils.py", line 31, in run_ddp
> > 0:     main_fn(args, rank)
> > 0:   File "/private/home/zhangguanheng/tmp/PyTorch/text/examples/BERT/cross_lingual_mlm_task.py", line 155, in run_main
> > 0:     self.register_comm_hook(state, powerSGD.powerSGD_hook)
> > 0: NameError: name 'self' is not defined
> > 0:
> > srun: error: learnfair1730: task 0: Exited with exit code 1
> > srun: launch/slurm: _step_signal: Terminating StepId=37541787.0
> > ```
> 
> Thanks! Fixed it. Please make sure you are using NCCL backend.

Yep. I use nccl backend. Do we want to merge this PR? I think we just need to add a `args` to allow the option to run PowerSGD and land this diff easily.

**Comment by SciPioneer** [comment]
I added all the args that can be used for tuning PowerSGD performance. Some features are only available after PyTorch 1.8.

Since I am not sure if we will use the 1.8 release or the master branch for experimentation, I commented out the code that involves the new features after 1.8.

File: examples/BERT/cross_lingual_mlm_task.py [hunk 1]
```diff
@@ -140,6 +142,40 @@ def text_transform(x: str) -> List:
         model = CrossLingualMLMTask(ntokens, args.emsize, args.nhead, args.nhid, args.nlayers)
         model = model.to(devices[0])
         model = DDP(model, device_ids=devices)
+
+        if args.comm_hook_type != `None`:
```
Comment (zhangguanheng66) [inline examples/BERT/cross_lingual_mlm_task.py:146]: `None` seems not a valid string. Do you want to use "None"?

File: examples/BERT/cross_lingual_mlm_task.py [hunk 2]
```diff
@@ -140,6 +142,40 @@ def text_transform(x: str) -> List:
...
+                # Defer the gradient compression until `start_powerSGD_iter` step.
+                start_powerSGD_iter=args.start_powerSGD_iter,
+                # Only apply PowerSGD to a layer if `min_compression_rate` can be achieved.
+                # This is not used by `BatchedPowerSGD` hook.
+                # WARNING: This feature is only available after PyTorch 1.8.
+                # min_compression_rate=args.min_compression_rate,
+            )
+            
+            hook = None
+            if args.comm_hook_type == `PowerSGD`:
```
Comment (zhangguanheng66) [inline examples/BERT/cross_lingual_mlm_task.py:163]: Same here for the quote mark.

File: examples/BERT/cross_lingual_mlm_task.py [hunk 3]
```diff
@@ -140,6 +142,40 @@ def text_transform(x: str) -> List:
...
+                hook = powerSGD.powerSGD_hook
+                model.register_comm_hook(state, )
+            elif args.comm_hook_type == `BatchedPowerSGD`:
```
Comment (zhangguanheng66) [inline examples/BERT/cross_lingual_mlm_task.py:166]: Same here for the quote mark.

**Comment by SciPioneer** [comment]
> LGTM. Could you also update the [README](https://github.com/pytorch/text/blob/xlmr_mlm/examples/BERT/README.md) file with a little bit more explanation plus an example command to run the PowerSGD. The DDP support was included at the bottom of the file.

Added some description, PTAL. Thanks!

**Pull request merged**